airflow/www/views.py

```diff
@@ -1738,10 +1738,21 @@ def gantt(self, session=None):
         gantt_bar_items = []
         for ti in tis:
             end_date = ti.end_date or timezone.utcnow()
-            gantt_bar_items.append((ti.task_id, ti.start_date, end_date, ti.state))
+            try_count = ti.try_number
+            if ti.state == State.FAILED or ti.state == State.SUCCESS:
+                try_count = ti.try_number - 1
+            gantt_bar_items.append((ti.task_id, ti.start_date, end_date, ti.state, try_count))
+
+        tf_count = 0
+        prev_task_id = ""
         for tf in ti_fails:
             end_date = tf.end_date or timezone.utcnow()
-            gantt_bar_items.append((tf.task_id, tf.start_date, end_date, State.FAILED))
+            try_count = 1
+            if tf_count != 0 and tf.task_id == prev_task_id:
+                try_count = try_count + 1
```

Comment:
Does this mean `try_count` can at most be 2?

Reply:
No, actually try_count was initialised out of the block. Now made the correction.

Reply:
If `ti_fails` is not sorted this piece of logic won't work correct, will it? If it is sorted, you still need to reset `try_count`. Also can you elaborate the difference of the bar items we are adding here versus the items we were adding [here](https://github.com/apache/airflow/pull/5037/files#diff-948e87b4f8f644b3ad8c7950958df033R1744) please?

Reply:
1. ti_fails will always be in sorted order because of the nature of task_fail table having auto increment key. 
2. try_count is always reset as 1 in the starting of the for loop of ti_fails. 
3. gantt_bar_items appended against the tis metadata are having the state as success or fail from task_instance table so having the try_number for the same only. In case of failed tasks, entries go into task_fail table and we can't get the try number from there directly hence I am incrementing the try_count and appending the same against ti_fails metadata because of the nature of task_fail table.

Reply:
1. I don't think having auto increment key has anything to do with the order of data returned. A quick search gave me [this](https://dba.stackexchange.com/questions/5774/why-is-ssms-inserting-new-rows-at-the-top-of-a-table-not-the-bottom/5775#5775https://dba.stackexchange.com/questions/5774/why-is-ssms-inserting-new-rows-at-the-top-of-a-table-not-the-bottom/5775#5775).
3. Assuming appending twice will not create two row for the same task instance in the graph( please confirm this), we are showing total try count only if the task instance is in FAILED state. This can be quite confusing. We need to probably revisit and decide what exactly we want to display and at least make sure it is consistent between task instances in different states.

Reply:
Answer to your questions -
1. I understand the fact but that order is already maintained through this ordered [list](https://github.com/apache/airflow/pull/5037/files#diff-948e87b4f8f644b3ad8c7950958df033R1728)  keeping the gantt chart in mind. Anyhow metadata apart from try number is coming in sorted order.
Example can be understood by the below use case -

mysql> select * from task_fail as tf  where TF.task_id = 'get_op'  AND TF.execution_date = '2019-05-29 05:30:00.000000' AND TF.dag_id='example_http_operator1';
+----+---------+------------------------+----------------------------+----------------------------+----------------------------+----------+
| id | task_id | dag_id                 | execution_date             | start_date                 | end_date                   | duration |
+----+---------+------------------------+----------------------------+----------------------------+----------------------------+----------+
|  8 | get_op  | example_http_operator1 | 2019-05-29 05:30:00.000000 | 2019-05-30 22:37:45.177828 | 2019-05-30 22:37:51.449197 |        6 |
| 27 | get_op  | example_http_operator1 | 2019-05-29 05:30:00.000000 | 2019-05-30 22:43:08.766188 | 2019-05-30 22:43:10.270735 |        2 |
| 46 | get_op  | example_http_operator1 | 2019-05-29 05:30:00.000000 | 2019-05-30 22:48:20.494280 | 2019-05-30 22:48:22.069887 |        2 |
| 62 | get_op  | example_http_operator1 | 2019-05-29 05:30:00.000000 | 2019-05-30 22:53:27.867508 | 2019-05-30 22:53:29.474945 |        2 |
+----+---------+------------------------+----------------------------+----------------------------+----------------------------+----------+
4 rows in set (0.00 sec)

mysql> select * from task_fail as tf  where TF.task_id = 'post_op'  AND TF.execution_date = '2019-05-29 05:30:00.000000' AND TF.dag_id='example_http_operator1';
+----+---------+------------------------+----------------------------+----------------------------+----------------------------+----------+
| id | task_id | dag_id                 | execution_date             | start_date                 | end_date                   | duration |
+----+---------+------------------------+----------------------------+----------------------------+----------------------------+----------+
| 17 | post_op | example_http_operator1 | 2019-05-29 05:30:00.000000 | 2019-05-30 22:38:19.100704 | 2019-05-30 22:38:26.619226 |        8 |
| 36 | post_op | example_http_operator1 | 2019-05-29 05:30:00.000000 | 2019-05-30 22:43:36.747317 | 2019-05-30 22:43:38.238442 |        1 |
| 53 | post_op | example_http_operator1 | 2019-05-29 05:30:00.000000 | 2019-05-30 22:48:48.267570 | 2019-05-30 22:48:49.595367 |        1 |
| 69 | post_op | example_http_operator1 | 2019-05-29 05:30:00.000000 | 2019-05-30 22:53:55.739041 | 2019-05-30 22:53:57.038769 |        1 |
+----+---------+------------------------+----------------------------+----------------------------+----------------------------+----------+
4 rows in set (0.00 sec)
Explanation : list has been formed by taking each individual task like appending the above result. Please find the below screenshot to validate the same results -
<img width="1560" alt="Screen Shot 2019-05-31 at 12 24 37 AM" src="https://user-images.githubusercontent.com/20847563/58656806-93a64500-833a-11e9-8df3-ab91d5ce1c65.png">

2. Yes, It is not appending twice. reasoning is same as above. All the metadata except try count is populated through the same logic. I have just added this only in both the gantt_bar_items. Above than that, I am showing the try count in every state. As task_fail doesn't have try number in it's table so I have made the logic to get the same and that's how try number is appended in gantt_bar_items of the failed tasks as well. 
Moreover, the conclusion is that we need to display try_count every-time user the Gantt UI of airflow irrespective of state and that is consistent across different states.

Reply:
1. Yes all task fails are grouped by task_id as we have only one ti for each task, missed this part earlier.  Tho we can still have wrong order of task fails. Result is that in the gantt chart, we can have a try with later start date appear on the left side of a try with earlier start date.
2. Mb misunderstood the state there, didn't get that append will create a new column in the row. In this case, did we append the last try for failed tasks twice?( visually doesn't seem to create a problem tho)

Reply:
Please find my answers -
1. Start date is already coming in the correct order even with out introducing my changes. I have only added the try number in every column. So gantt chart will not have any case in which later start date appears on the left side as compare to earlier start date.
2. Even in this, append is already creating the column twice with out introducing my changes and yes there is no problem with this as well.

Reply:
The value of [ti_fails](https://github.com/apache/airflow/pull/5037/files#diff-948e87b4f8f644b3ad8c7950958df033R1728) is not guaranteed to be ordered but you don't have to include a fix in your PR as it is not newly introduced. I'll submit a new JIRA for it.